float64[3] n
float64[3] o
float64[3] a
float64[3] p
